### DISCUSSION TITLE
Checking current state (reducer) using standard 'set' updates

### DISCUSSION THREAD

**Discussion by gavinthomas-valtech** [pmndrs/zustand#2552] [Q&A] [answered]
I very much like reducer patterns which enable me to check states and conditions before setting states. This enables a more state machine like pattern. What i wanted to know can you do this using the standard `set` pattern?
eg

```
const usePersonStore = create<State & Action>((set) => ({
  firstName: '',
  lastName: '',
  updateFirstName: (firstName) => set(() => ({ firstName: firstName })),
  updateLastName: (lastName) => set(() => ({ lastName: lastName })),
}))

```

**Comment by dbritto-dev** [comment]
@gavinthomas-valtech yes,`setState` function supports callbacks

**Reply by gavinthomas-valtech** [reply]
HOw does that look

for example i often do this

```
 case 'RESET_STATE':
      if (state.viewState === 'activeSearch') {
        return { ...InitialState };
      }
```

in a reducer pattern. Or switch on state
```
 switch (state.viewState) {
    case viewStates.initialChecks: {
      if (event.type === 'PREPARE_TO_REJECT') {
        return {
```

This enables me to centralise logic and only send events from components. How does this look in the store?

**Reply by dbritto-dev** [reply, marked as answer]
```ts
set(currentState => ({ ... })
```

**Reply by dbritto-dev** [reply]
but as far I see, seems you want to use redux. btw, there's redux middleware for redux-like implementations

**Reply by gavinthomas-valtech** [reply]
indeed. I do prefer having actions separated from where the store is created. I'm currently working in a codebase that uses the standard zustand store set up but i've written a lot of local (useReducer) logic that now needs to live in the globsal store
